### PR TITLE
feat(graphs): Enable multi-level lam resolution in stretched grid with cutout mask #1115

### PR DESCRIPTION
**Pull request by jasonjunweilyu**
## Description

This PR enables stretched-grid graph to have multi-level lam resolutions. A function is added to create multi stretched-grid tri nodes. Stretched grid node builder and multi-scale edges are updated for this feature. Tests are added for the stretched grid node builder and updated for multi-scale edges.

## What problem does this change solve?

New feature.

## What issue or task does this change relate to?

#1115 

##  Additional notes ##

See example multi-resolution lam region test config, log and plots in the comment section below.
For single-resolution lam stretched-grid, this PR is backward compatible, so no change to the previous config. 

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

### CODE REVIEW

**Comment by jasonjunweilyu** [comment]
This was tested with the following config and produced the log and plots as shown below:

<details>
  <summary>Config: </summary>

```yaml
overwrite: True

nodes:
  data:
    node_builder:
      _target_: anemoi.graphs.nodes.AnemoiDatasetNodes
      dataset: 
        cutout:
          # use the same BARRA datasets since I current only has BARRA_C (4.4km), not BARRA_C (12km)
          - dataset: /g/data/ma14/anemoi_training/dataset/barra/test_barra2_01-02-2020_1216day.zarr
            area: [-10.0, 120.0, -30.0, 140.0]
            reorder: sort
          - dataset: /g/data/ma14/anemoi_training/dataset/barra/test_barra2_01-02-2020_1216day.zarr
            area: [-4.01, 107.02, -46.69, 160.9]
            reorder: sort
          - join:
            - dataset: /g/data/dx2/hc7988/ECMWF/Datasets/aifs-od-an-oper-0001-mars-n320-2016-2023-6h-v8.zarr
              drop: ["sdor", "slor", "cp", "u_600", "v_600", "z_600", "t_600", "q_600", "w_600"]
            - dataset: /g/data/dx2/hc7988/ECMWF/Datasets/aifs-od-an-oper-0001-mars-n320-2016-2023-6h-v1-land.zarr
              select: ["lcc", "mcc", "hcc", "tcc", "strd", "ssrd"]
            adjust: ["start", "end"]
            reorder: sort
        min_distance_km: 0
        adjust: all
    attributes: 
      barra3_mask:
        _target_: anemoi.graphs.nodes.attributes.GridsMask
        grids: 0
      barra2_mask:
        _target_: anemoi.graphs.nodes.attributes.GridsMask
        grids: 1
      boundary_mask:
        _target_: anemoi.graphs.nodes.attributes.GridsMask
        grids: 2
      lam_mask:
        _target_: anemoi.graphs.nodes.attributes.BooleanNot
        masks: boundary_mask
      area_weight:
        _target_: anemoi.graphs.nodes.attributes.SphericalAreaWeights
        norm: unit-max
        fill_value: 0
      lam_area_weight:
        _target_: anemoi.graphs.nodes.attributes.MaskedPlanarAreaWeights
        mask_node_attr_name: lam_mask
        norm: unit-max

  hidden:
    node_builder:
      _target_: anemoi.graphs.nodes.StretchedTriNodes
      global_resolution: 5
      lam_regions:
        - lam_resolution: 7
          reference_node_name: data
          mask_attr_name: barra3_mask
          margin_radius_km: 11
        - lam_resolution: 6
          reference_node_name: data
          mask_attr_name: barra2_mask
          margin_radius_km: 11

edges:
# Encoder
- source_name: data
  target_name: hidden
  edge_builders:
  - _target_: anemoi.graphs.edges.KNNEdges
    num_nearest_neighbours: 12
    source_mask_attr_name: null
    target_mask_attr_name: null
  attributes:
    edge_length:
      _target_: anemoi.graphs.edges.attributes.EdgeLength
      norm: unit-max
    edge_dirs:
      _target_: anemoi.graphs.edges.attributes.EdgeDirection
      norm: unit-std
# Processor: update scale_resolutions
- source_name: hidden
  target_name: hidden
  edge_builders:
  - _target_: anemoi.graphs.edges.MultiScaleEdges
    x_hops: 1
    scale_resolutions: 7
    source_mask_attr_name: null
    target_mask_attr_name: null
  attributes: 
    edge_length:
      _target_: anemoi.graphs.edges.attributes.EdgeLength
      norm: unit-max
    edge_dirs:
      _target_: anemoi.graphs.edges.attributes.EdgeDirection
      norm: unit-std
# Decoder
- source_name: hidden
  target_name: data
  edge_builders:
  - _target_: anemoi.graphs.edges.KNNEdges
    num_nearest_neighbours: 3
    source_mask_attr_name: null
    target_mask_attr_name: null
  attributes: 
    edge_length:
      _target_: anemoi.graphs.edges.attributes.EdgeLength
      norm: unit-max
    edge_dirs:
      _target_: anemoi.graphs.edges.attributes.EdgeDirection
      norm: unit-std

post_processors: []

```

</details>


<details>
  <summary>Log:</summary>

  ```
2026-05-18 14:34:19 INFO Reading the dataset from {'cutout': [{'dataset': '/g/data/ma14/anemoi_training/dataset/barra/test_barra2_01-02-2020_1216day.zarr', 'area': [-10.0, 120.0, -30.0, 140.0], 'reorder': 'sort'}, {'dataset': '/g/data/ma14/anemoi_training/dataset/barra/test_barra2_01-02-2020_1216day.zarr', 'area': [-4.01, 107.02, -46.69, 160.9], 'reorder': 'sort'}, {'join': [{'dataset': '/g/data/dx2/hc7988/ECMWF/Datasets/aifs-od-an-oper-0001-mars-n320-2016-2023-6h-v8.zarr', 'drop': ['sdor', 'slor', 'cp', 'u_600', 'v_600', 'z_600', 't_600', 'q_600', 'w_600']}, {'dataset': '/g/data/dx2/hc7988/ECMWF/Datasets/aifs-od-an-oper-0001-mars-n320-2016-2023-6h-v1-land.zarr', 'select': ['lcc', 'mcc', 'hcc', 'tcc', 'strd', 'ssrd']}], 'adjust': ['start', 'end'], 'reorder': 'sort'}], 'min_distance_km': 0, 'adjust': 'all'}.
2026-05-18 14:34:20 DEBUG Registering codec 'zlib'
2026-05-18 14:34:20 DEBUG Registering codec 'gzip'
2026-05-18 14:34:20 DEBUG Registering codec 'bz2'
2026-05-18 14:34:20 DEBUG Registering codec 'lzma'
2026-05-18 14:34:20 DEBUG Registering codec 'blosc'
2026-05-18 14:34:20 DEBUG Registering codec 'zstd'
2026-05-18 14:34:20 DEBUG Registering codec 'lz4'
2026-05-18 14:34:20 DEBUG Registering codec 'astype'
2026-05-18 14:34:20 DEBUG Registering codec 'delta'
2026-05-18 14:34:20 DEBUG Registering codec 'quantize'
2026-05-18 14:34:20 DEBUG Registering codec 'fixedscaleoffset'
2026-05-18 14:34:20 DEBUG Registering codec 'packbits'
2026-05-18 14:34:20 DEBUG Registering codec 'categorize'
2026-05-18 14:34:20 DEBUG Registering codec 'pickle'
2026-05-18 14:34:20 DEBUG Registering codec 'base64'
2026-05-18 14:34:20 DEBUG Registering codec 'shuffle'
2026-05-18 14:34:20 DEBUG Registering codec 'bitround'
2026-05-18 14:34:20 DEBUG Registering codec 'crc32'
2026-05-18 14:34:20 DEBUG Registering codec 'adler32'
2026-05-18 14:34:20 DEBUG Registering codec 'jenkins_lookup3'
2026-05-18 14:34:20 DEBUG Registering codec 'json2'
2026-05-18 14:34:20 DEBUG Registering codec 'vlen-utf8'
2026-05-18 14:34:20 DEBUG Registering codec 'vlen-bytes'
2026-05-18 14:34:20 DEBUG Registering codec 'vlen-array'
2026-05-18 14:34:20 DEBUG Registering codec 'fletcher32'
2026-05-18 14:34:20 DEBUG Registering codec 'n5_wrapper'
2026-05-18 14:35:28 DEBUG Time to register node coordinates (AnemoiDatasetNodes): 68.04 s
2026-05-18 14:36:29 DEBUG GridsMask values are not normalised.
2026-05-18 14:37:31 DEBUG GridsMask values are not normalised.
2026-05-18 14:38:33 DEBUG GridsMask values are not normalised.
2026-05-18 14:38:33 DEBUG BooleanNot values are not normalised.
2026-05-18 14:39:13 DEBUG There are 1839044 of weights, which (unscaled) add up a total weight of 12.57.
2026-05-18 14:41:58 DEBUG Time to register node coordinates (AnemoiDatasetNodes): 390.81 s
2026-05-18 14:41:59 INFO Fitting KNNAreaMaskBuilder with 250000 reference nodes from "data (barra3_mask)".
2026-05-18 14:41:59 INFO Fitting KNNAreaMaskBuilder with 1071364 reference nodes from "data (barra2_mask)".
2026-05-18 14:42:02 DEBUG Time to register node coordinates (StretchedTriNodes): 3.84 s
2026-05-18 14:42:02 DEBUG Time to register node coordinates (StretchedTriNodes): 0.00 s
2026-05-18 14:42:02 INFO Using KNN-Edges (with 12 nearest neighbours) between data and hidden.
2026-05-18 14:42:02 WARNING The 'torch-cluster' library is not installed.
Installing 'torch-cluster' can significantly improve performance for graph creation.
You can install it using:
    TORCH_VERSION=$(python -c "import torch; print(torch.__version__)")
    pip install torch-cluster -f https://data.pyg.org/whl/torch-${TORCH_VERSION}.html

2026-05-18 14:42:03 DEBUG Time to register edge indices (KNNEdges): 1.10 s
2026-05-18 14:42:03 DEBUG Adding 240 edges for resolution 1
2026-05-18 14:42:04 DEBUG Adding 960 edges for resolution 2
2026-05-18 14:42:04 DEBUG Adding 3840 edges for resolution 3
2026-05-18 14:42:04 DEBUG Adding 15360 edges for resolution 4
2026-05-18 14:42:06 DEBUG Adding 61440 edges for resolution 5
2026-05-18 14:42:07 DEBUG Adding 10936 edges for resolution 6
2026-05-18 14:42:12 DEBUG Adding 8930 edges for resolution 7
2026-05-18 14:42:12 DEBUG Time to register edge indices (MultiScaleEdges): 8.25 s
2026-05-18 14:42:12 INFO Using KNN-Edges (with 3 nearest neighbours) between hidden and data.
2026-05-18 14:42:12 WARNING The 'torch-cluster' library is not installed.
Installing 'torch-cluster' can significantly improve performance for graph creation.
You can install it using:
    TORCH_VERSION=$(python -c "import torch; print(torch.__version__)")
    pip install torch-cluster -f https://data.pyg.org/whl/torch-${TORCH_VERSION}.html

2026-05-18 14:42:16 DEBUG Time to register edge indices (KNNEdges): 3.99 s
2026-05-18 14:42:17 INFO Cleaning graph.
2026-05-18 14:42:17 INFO _grid_reference_distance deleted from graph.
2026-05-18 14:42:17 INFO _dataset deleted from graph.
2026-05-18 14:42:17 INFO _grid_reference_distance deleted from graph.
2026-05-18 14:42:17 INFO _nx_graph deleted from graph.
2026-05-18 14:42:17 INFO _resolutions deleted from graph.
2026-05-18 14:42:17 INFO _node_ordering deleted from graph.
2026-05-18 14:42:17 INFO Graph saved at graph_new-debug.pt.

📦 Path       : graph_new-debug.pt
💽 Size       : 145.3 MiB (152323856)

🪩  Nodes summary

   Nodes name │  Num. nodes │                                                                      Attributes │ Attribute dim │ Min. latitude │ Max. latitude │ Min. longitude │ Max. longitude
   ───────────┼─────────────┼─────────────────────────────────────────────────────────────────────────────────┼───────────────┼───────────────┼───────────────┼────────────────┼───────────────
   data       │ 1.83904e+06 │ barra3_mask, barra2_mask, boundary_mask, lam_mask, area_weight, lam_area_weight │             6 │      -89.7849 │       89.7849 │              0 │        359.719
   hidden     │       12767 │                                                                                 │             0 │           -90 │            90 │       -178.606 │            180
   ───────────┴─────────────┴─────────────────────────────────────────────────────────────────────────────────┴───────────────┴───────────────┴───────────────┴────────────────┴───────────────


🌐  Edges summary

   Source │ Target │  Num. edges │ Isolated Source │ Isolated Target │ Min edges/target │ Max edges/target │ Attribute dim │                     Attributes
   ───────┼────────┼─────────────┼─────────────────┼─────────────────┼──────────────────┼──────────────────┼───────────────┼───────────────────────────────
   data   │ hidden │      153204 │     1.68584e+06 │               0 │               12 │               12 │             3 │ edge_length(1D), edge_dirs(2D)
   hidden │ hidden │      101706 │               0 │               0 │                3 │               36 │             3 │ edge_length(1D), edge_dirs(2D)
   hidden │ data   │ 5.51713e+06 │               0 │               0 │                3 │                3 │             3 │ edge_length(1D), edge_dirs(2D)
   ───────┴────────┴─────────────┴─────────────────┴─────────────────┴──────────────────┴──────────────────┴───────────────┴───────────────────────────────


📊 Attribute distributions

   Type │ Source          │            Name │         Dtype │        Min. │         Mean │    Max. │  Std. dev.
   ─────┼─────────────────┼─────────────────┼───────────────┼─────────────┼──────────────┼─────────┼───────────
   Node │ data            │     barra3_mask │    torch.bool │           0 │      0.13594 │       1 │   0.342725
   Node │ data            │     barra2_mask │    torch.bool │           0 │     0.582566 │       1 │   0.493136
   Node │ data            │   boundary_mask │    torch.bool │           0 │     0.281494 │       1 │   0.449728
   Node │ data            │        lam_mask │    torch.bool │           0 │     0.718506 │       1 │   0.449728
   Node │ data            │     area_weight │ torch.float32 │   0.0114431 │     0.283801 │       1 │   0.424887
   Node │ data            │ lam_area_weight │ torch.float32 │           0 │   0.00647255 │       1 │ 0.00849534
   Edge │ data-->hidden   │     edge_length │ torch.float32 │ 0.000663291 │     0.490239 │       1 │   0.295139
   Edge │ data-->hidden   │       edge_dirs │ torch.float32 │    -1.41421 │ -3.64397e-05 │ 1.41421 │          1
   Edge │ hidden-->hidden │     edge_length │ torch.float32 │   0.0137661 │    0.0748275 │       1 │  0.0731556
   Edge │ hidden-->hidden │       edge_dirs │ torch.float32 │    -1.41421 │  3.41072e-05 │ 1.41421 │          1
   Edge │ hidden-->data   │     edge_length │ torch.float32 │ 0.000165083 │     0.340407 │       1 │   0.197065
   Edge │ hidden-->data   │       edge_dirs │ torch.float32 │    -1.41421 │  0.000305964 │ 1.41421 │          1
   ─────┴─────────────────┴─────────────────┴───────────────┴─────────────┴──────────────┴─────────┴───────────

🔋 Graph ready.

  ```

</details>


<details>
  <summary>Plots: </summary>

Hidden layer nodes:

image

Zoomed-in hidden to hidden:

image

Zoomed-in data to hidden:

image

</details>

**Comment by JPXKQX** [comment]
Thanks, @jasonjunweilyu, for the PR and the detailed example config and plots. It's great to see the multi-resolution stretched-grid working end-to-end.

Before we can merge this feature, we'd like to see evidence of added scientific value compared to the existing single-resolution stretched-grid baseline. The extra complexity introduced by new features carries a maintenance cost, so the Anemoi Technical Subgroup agreed to verify that the new features provide added value (e.g. improved forecast quality in this case).

Following your example in Issue #1115 , I would compare the proposed multi-level stretched grid against:
- `lam_resolution: 8` (in Australia) and `global_resolution: 5` (for the rest of the earth)
- `lam_resolution: 6` (in Australasia) and `global_resolution: 5` (for the rest of the earth)

Thanks again for putting the time into expanding anemoi with new features!

**Comment by jasonjunweilyu** [comment]
Hi @JPXKQX , thanks for considering this feature and doing the review. I agree that the efforts would not be trivial to maintain this feature since if it works for stretched-grid, you may also want it for lam and also other type of grids. 

I will report back here in a month's time when I get the comparison results, hopefully. Right now, we are still fixing the BARRA training dataset to align it better with IFS training dataset. We will resume training once this is done.
